File: backend/farm2route-api/src/main/java/com/farm2route/agency/dto/AgencySignupRequest.java
package com.farm2route.agency.dto;

import com.farm2route.agency.enums.AgencyType;
import com.farm2route.agency.validation.PasswordMatch;
import com.farm2route.agency.validation.ValidBusinessRegNumber;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@PasswordMatch
@ValidBusinessRegNumber
public class AgencySignupRequest {

    @NotBlank(message = "Agency name is required")
    private String agencyName;

    @NotBlank(message = "Email is required")
    @Email(message = "Invalid email format")
    private String email;

    @NotBlank(message = "Phone number is required")
    private String phoneNumber;

    @NotBlank(message = "Password is required")
    @Size(min = 8, message = "Password must be at least 8 characters")
    private String password;

    @NotBlank(message = "Confirm password is required")
    private String confirmPassword;

    @NotNull(message = "Agency type is required")
    private AgencyType agencyType;

    private String businessRegNumber;

    @NotBlank(message = "District is required")
    private String district;

    @NotBlank(message = "Address is required")
    private String address;

    private String contactPersonName;
}

File: mobile/farm2route_app/lib/features/agency/data/models/agency_signup_request.dart
class AgencySignupRequest {
  final String agencyName;
  final String email;
  final String phoneNumber;
  final String password;
  final String confirmPassword;
  final String agencyType;
  final String? businessRegNumber;
  final String district;
  final String address;
  final String? contactPersonName;

  const AgencySignupRequest({
    required this.agencyName,
    required this.email,
    required this.phoneNumber,
    required this.password,
    required this.confirmPassword,
    required this.agencyType,
    this.businessRegNumber,
    required this.district,
    required this.address,
    this.contactPersonName,
  });

  Map<String, dynamic> toJson() {
    return {
      'agencyName': agencyName,
      'email': email,
      'phoneNumber': phoneNumber,
      'password': password,
      'confirmPassword': confirmPassword,
      'agencyType': agencyType,
      if (businessRegNumber != null && businessRegNumber!.trim().isNotEmpty)
        'businessRegNumber': businessRegNumber!.trim(),
      'district': district,
      'address': address,
      if (contactPersonName != null && contactPersonName!.trim().isNotEmpty)
        'contactPersonName': contactPersonName!.trim(),
    };
  }

  factory AgencySignupRequest.fromJson(Map<String, dynamic> json) {
    return AgencySignupRequest(
      agencyName: json['agencyName'] as String,
      email: json['email'] as String,
      phoneNumber: json['phoneNumber'] as String,
      password: json['password'] as String,
      confirmPassword: json['confirmPassword'] as String,
      agencyType: json['agencyType'] as String,
      businessRegNumber: json['businessRegNumber'] as String?,
      district: json['district'] as String,
      address: json['address'] as String,
      contactPersonName: json['contactPersonName'] as String?,
    );
  }
}
